crates/cxx-qt-build/src/lib.rs

```diff
@@ -536,110 +555,129 @@ impl CxxQtBuilder {
+        // MSVC
+        builder.flag_if_supported("/Zc:__cplusplus");
+        builder.flag_if_supported("/permissive-");
+        builder.flag_if_supported("/bigobj");
+        // MinGW requires big-obj otherwise debug builds fail
+        builder.flag_if_supported("-Wa,-mbig-obj");
+
+        // Enable any extra defines
+        for define in defines {
+            builder.define(define, None);
         }
 
-        for builder in [&mut self.cc_builder, &mut cc_builder_whole_archive] {
-            // Note, ensure our settings stay in sync across cxx-qt, cxx-qt-build, and cxx-qt-lib
-            builder.cpp(true);
-            builder.std("c++17");
-            // MSVC
-            builder.flag_if_supported("/Zc:__cplusplus");
-            builder.flag_if_supported("/permissive-");
-            builder.flag_if_supported("/bigobj");
-            // MinGW requires big-obj otherwise debug builds fail
-            builder.flag_if_supported("-Wa,-mbig-obj");
-
-            // Enable any extra defines
-            for extra_define in &self.extra_defines {
-                builder.define(extra_define, None);
-            }
+        for include_path in include_paths {
+            builder.include(include_path);
+        }
+    }
 
-            builder.includes(qtbuild.include_paths());
-            builder.include(&header_root);
-            builder.include(&generated_header_dir);
+    fn moc_qobject_headers(&mut self, qtbuild: &mut qt_build_utils::QtBuild) {
+        for QObjectHeaderOpts {
+            path,
+            moc_arguments,
+        } in &self.qobject_headers
+        {
+            let moc_products = qtbuild.moc(path, moc_arguments.clone());
+            self.cc_builder.file(moc_products.cpp);
         }
+    }
 
-        // Generate files
-        for files in generate_cxxqt_cpp_files(&self.rust_sources, &generated_header_dir) {
+    fn generate_cpp_files_from_cxxqt_bridges(&mut self, header_dir: impl AsRef<Path>) {
+        for files in generate_cxxqt_cpp_files(&self.rust_sources, &header_dir) {
             self.cc_builder.file(files.plain_cpp);
             if let (Some(qobject), Some(qobject_header)) = (files.qobject, files.qobject_header) {
                 self.cc_builder.file(&qobject);
                 self.qobject_headers.push(qobject_header.into());
             }
         }
+    }
 
-        // Run moc on C++ headers with Q_OBJECT macro
-        for QObjectHeaderOpts {
-            path,
-            moc_arguments,
-        } in self.qobject_headers
-        {
-            let moc_products = qtbuild.moc(&path, moc_arguments);
-            self.cc_builder.file(moc_products.cpp);
+    fn build_object_file(
+        builder: &cc::Build,
+        file_path: impl AsRef<Path>,
+        export_path: Option<(&str, &str)>,
+    ) {
+        let mut obj_builder = builder.clone();
+        obj_builder.file(file_path);
+        let obj_files = obj_builder.compile_intermediates();
+
+        if let [obj_file] = &obj_files[..] {
```

Comment:
is this taking the first element? if so why not

```rust
if let Some(obj_file) = obj_files.first() {
...
}
```

Reply:
Because this way we get to error out if there are no, or even more than one obj files as the output.
It should never be that there's more than one obj file produced by the builder, so I want to error out in that case.

I guess I could also check the `len`, but I think this is quite elegant, because we're confirming that the vec is structured as expected and accessing the right element at the same time.